### ISSUE TITLE
SignedXml. Ability to call SignedInfo verification or improve return type.

### ISSUE THREAD

**Issue by isakgjaerum** [dotnet/runtime#73959]
We are currently working with the SignedXml class and have come across several difficulties with it.
- First of all the SignedXml.CheckSignature method performs four different checks while the return type is a boolean leaving us with no clue as to which of the checks went wrong when we get a false return.
- Secondly, we are experiencing that the XmlDocument class is tampering with the contents of the XML, and since the SignedXml class only accepts XmlDocument or XmlElement in its constructors the XML is never verified. Although the SignedInfo verifies, which we found out through stepping into the code. As to what the XmlDocument is tampering with, we have some XMLs containing line breaks (&# 13;) that are turned into carriage return line feed, leaving the XML in a different state than when it was signed. 

We see several things that would make the use of SignedXml better, first and foremost the ability to differentiate between what went wrong in the CheckSignature method by either returning a more detailed message. 

Furthermore, make it possible for us to call each of those methods (CheckSignatureFormat, CheckSignedInfo, and CheckDigestedReferences) separately so that we know exactly which of them went wrong. Arguably an easier solution than the above one.

Another solution would be to allow us to call SignedXml with a Stream instead of an XML. We are working with enveloped detached signatures with non-resolvable URI's so we do the process of checking the references ourselves. This also allows us to call the C14N transformations with a MemoryStream which does not tamper with the XML. We are then using the SignedXml class to find the algorithms and transforms needed but use them on the Stream that has not been tampered with by XmlDocument. 

To sum it up, the minimal viable solution that would satisfy us would be to make a public method on the SignedXml class allowing us to call only CheckSignedInfo. We will be happy to assist you.

**Comment by vcsjones** [comment]
A related API proposal: https://github.com/dotnet/runtime/issues/27263

**Comment by bartonjs** [comment]
Just to set some expectations: we're not really investing in SignedXml, other than fixing whatever security bugs we can (many of them start at the spec and are therefore hard to deal with).

My most sincere advise is don't use SignedXml at all, or any other component that uses the xmldsig specification, unless you're forced to by interoperating with an existing system/partner.

**Comment by isakgjaerum** [comment]
Yeah we are definitely in the last category, we are aware of the drawbacks of this, but are forced due to interoperating with existing systems. 

I figured this was an area with little to no imvestment. I do though see the minimal solution i suggested as a quick and easy thing to implement. Would it be betrer to submit a PR with a proposal?

**Comment by bartonjs** [comment]
> Would it be better to submit a PR with a proposal?

Starting with a PR is discouraged, because any changes to public API have to go through API review.  Starting with a prototype (to make sure that it can work the-way/for-what you want it to), turning that into an API proposal, and then after that gets discussed and approved then turning that into a PR would be OK.

One thing that I worry about with making things more piecemeal is that it would increase the room for confusion... e.g. did someone check only that the SignedInfo was OK but didn't check any of the references, or checked that the reference hashes were sound but never that they weren't tampered with.  The type is already not super easy to use right (and sometimes it's too easy to use wrong, such as calling zero-argument `CheckSignature()`), and I'm wary that more knobs would make it even worse.

Something like needing to call `signedXml.SignedInfo.CheckLocalSignature(AsymmetricAlgorithm)` or `signedXml.SignedInfo.CheckLocalSignature(HMAC)` ("Local" being a word I threw in to suggest it's not looking at References) and then (depending on where the data is layered right now) something like `signedXml.SignedInfo.References[N].CheckDigest()` would potentially be reasonable... not as in-your-face as putting them directly on SignedXml.
